minisat/mtl/Vec.h

```diff
@@ -36,93 +39,218 @@ namespace Minisat {
-    void     shrink_  (int nelems)     { assert(nelems <= sz); sz -= nelems; }
-    int      capacity (void) const     { return cap; }
+    int size() const { return m_current - m_data; }
+    bool empty() const { return m_current == m_data; }
+    void     shrink   (int nelems) {
+        assert(nelems <= size() && "Attempted to shrink vector by more than it has elements");
+        shrink_(nelems, std::is_trivially_destructible<T>{});
+    }
+    // Truncate the vector from position till the `end()`
+    void truncate(const_iterator from) {
+        // By using `std::less_equal` instead of `<=`, we have guaranteed
+        // the definedness of the comparison even if the iterators (pointers)
+        // are from a different allocation.
+        assert(std::less_equal<const_iterator>{}(m_data, from) && "Outside of vector");
+        assert(std::less_equal<const_iterator>{}(from, m_current) && "Outside of valid elements");
+        erase_(from, std::is_trivially_destructible<T>{});
+        m_current = const_cast<T*>(from);
+    }
+    int      capacity () const     { return m_end - m_data; }
     void     capacity (int min_cap);
     void     growTo   (int size);
     void     growTo   (int size, const T& pad);
     void     clear    (bool dealloc = false);
 
     // Stack interface:
-    void     push  (void)              { if (sz == cap) capacity(sz+1); new (&data[sz]) T(); sz++; }
-    void     push  (const T& elem)     { if (sz == cap) capacity(sz+1); data[sz++] = elem; }
-    void     push_ (const T& elem)     { assert(sz < cap); data[sz++] = elem; }
-    void     pop   (void)              { assert(sz > 0); sz--, data[sz].~T(); }
+    void push() {
+        if (m_current == m_end) {
+            capacity(size() + 1);
+        }
+        new (m_current) T();
+        ++m_current;
+    }
+    void push (const T& elem) {
+        if (m_current == m_end) {
+            capacity(size() + 1);
+        }
+        *m_current = elem;
+        ++m_current;
+    }
+    // Does not realloc if there is not enough space for the element!
+    void     push_ (const T& elem)     { assert(m_current != m_end); *(m_current++) = elem; }
+    void     pop() { assert(m_data != m_current && "Attempted vec::pop on an empty vec"); pop_(std::is_trivially_destructible<T>{}); }
     // NOTE: it seems possible that overflow can happen in the 'sz+1' expression of 'push()', but
     // in fact it can not since it requires that 'cap' is equal to INT_MAX. This in turn can not
     // happen given the way capacities are calculated (below). Essentially, all capacities are
     // even, but INT_MAX is odd.
 
-    const T& last  (void) const        { return data[sz-1]; }
-    T&       last  (void)              { return data[sz-1]; }
+    const T& last  () const        { return *(m_current - 1); }
+    T&       last  ()              { return *(m_current - 1); }
```

Comment:
The `assert` checks are a bit inconsistent. If `truncate(...)` has a thorough check (which I completely support), shall a similar check be also added to the `last()` and other similar methods?

Something like `assert(m_current != m_data);`?

Reply:
I left these as they were except for syntactic translation to the new format, but I agree that there is no reason not to add the assert.